### PR TITLE
Retained `Gizmo`s

### PR DESCRIPTION
**Pull request by tim-blackbird**
# Objective
Add a way to use the gizmo API in a retained manner, for increased performance.

## Solution
- Move gizmo API from `Gizmos` to `GizmoBuffer`, ~ab~using `Deref` to keep usage the same as before.
- Merge non-strip and strip variant of `LineGizmo` into one, storing the data in a `GizmoBuffer` to have the same API for retained `LineGizmo`s.

### Review guide
- The meat of the changes are in `lib.rs`, `retained.rs`, `gizmos.rs`, `pipeline_3d.rs` and `pipeline_2d.rs`
- The other files contain almost exclusively the churn from moving the gizmo API from `Gizmos` to `GizmoBuffer`

### Performance

Performance compared to the immediate mode API is from 65 to 80 times better for static lines.

```
7900 XTX, 3700X
1707.9k lines/ms: gizmos_retained (21.3ms)
3488.5k lines/ms: gizmos_retained_continuous_polyline (31.3ms)
   0.5k lines/ms: gizmos_retained_separate (97.7ms)

3054.9k lines/ms: bevy_polyline_retained_nan (16.8ms)
3596.3k lines/ms: bevy_polyline_retained_continuous_polyline (14.2ms)
   0.6k lines/ms: bevy_polyline_retained_separate (78.9ms)

  26.9k lines/ms: gizmos_immediate (14.9ms)
  43.8k lines/ms: gizmos_immediate_continuous_polyline (18.3ms)
```
Looks like performance is good enough, being close to par with `bevy_polyline`.

Benchmarks can be found here: 
This branch: https://github.com/tim-blackbird/line_racing/tree/retained-gizmos
Bevy 0.14: https://github.com/DGriffin91/line_racing

## Showcase
```rust 
fn setup(
    mut commands: Commands,
    mut gizmo_assets: ResMut<Assets<GizmoAsset>>
) {
    let mut gizmo = GizmoAsset::default();

    // A sphere made out of one million lines!
    gizmo
        .sphere(default(), 1., CRIMSON)
        .resolution(1_000_000 / 3);

    commands.spawn(Gizmo {
        handle: gizmo_assets.add(gizmo),
        ..default()
    });
}
```

## Follow-up work
- Port over to the retained rendering world proper
- Calculate visibility and cull `Gizmo`s

### CODE REVIEW

**Comment by tim-blackbird** [comment]
It's still a bit rough, but it's now in a working state!
Performance feels good, I'm going to plug it into https://github.com/DGriffin91/line_racing/ and have a look at some real numbers!

Feel free to try it out, but I'd hold off on reviews for now :)

File: crates/bevy_dev_tools/src/ui_debug_overlay/inset.rs [hunk 1]
```diff
@@ -184,9 +184,9 @@ impl<'w, 's> InsetGizmo<'w, 's> {
                 Vec2::new(right, bottom),
                 Vec2::new(right, top),
                 Vec2::new(left, top),
-            ];
-            self.draw
-                .linestrip_2d(strip.map(|v| self.relative(v)), color);
+            ]
+            .map(|v| self.relative(v));
+            self.draw.linestrip_2d(strip, color);
```
Comment (tim-blackbird) [inline crates/bevy_dev_tools/src/ui_debug_overlay/inset.rs:189]: This shows an instance where the use of `DerefMut` can cause compilation errors.
This issue of partial borrows is not exclusive `DerefMut` though, hopefully rust can get be smarter about this in the future with polonius.

**Comment by pablo-lua** [comment]
Hey there, I'm trying to test this locally, but no lines are showing. Maybe I'm doing something wrong, or is this not ready yet?

File: crates/bevy_gizmos/src/lib.rs [hunk 2]
```diff
@@ -476,59 +463,92 @@ struct LineGizmoUniform {
...
-    pub positions: Vec<Vec3>,
-    /// Colors of the gizmo's vertices
-    pub colors: Vec<LinearRgba>,
-    /// Whether this gizmo's topology is a line-strip or line-list
-    pub strip: bool,
-    /// Whether this gizmo should draw line joints. This is only applicable if the gizmo's topology is line-strip.
-    pub joints: GizmoLineJoint,
-    /// The type of the gizmo's configuration group
-    pub config_ty: TypeId,
+pub struct LineGizmoAsset {
```
Comment (pablo-lua) [inline crates/bevy_gizmos/src/lib.rs:470]: I'm wondering: What is the best way to change a entity retained gizmos? Like position for example
The only way now is by replacing this handle in the LineGizmos component, right?

File: examples/gizmos/3d_gizmos.rs [hunk 3]
```diff
@@ -18,13 +18,33 @@ fn main() {
...
-struct MyRoundGizmos {}
+struct MyRoundGizmos;
 
 fn setup(
     mut commands: Commands,
+    mut linegizmos: ResMut<Assets<LineGizmoAsset>>,
     mut meshes: ResMut<Assets<Mesh>>,
     mut materials: ResMut<Assets<StandardMaterial>>,
 ) {
+    let mut linegizmo = LineGizmoAsset::new();
```
Comment (pablo-lua) [inline examples/gizmos/3d_gizmos.rs:29]: Would be good to have this example showing how to despawn or move this line.

Comment (tim-blackbird) [reply examples/gizmos/3d_gizmos.rs:29]: You can move it around and despawn it like any entity, but I do think it's worth showing off how to do that
However, this example is already very busy visually. I'd like to clean it up a bit before putting more effort into it. Ideally in a separate PR.

**Comment by tim-blackbird** [comment]
I've fixed some issues with this PR.
- I botched the port to the retained render world initially. I've re-done the port with bare minimum effort using `TemporaryRenderEntity` to keep things simple here.
- I initially had used the whole `GizmoConfig` in the `LineGizmo` component, but I overlooked the render layers setting in there. `RenderLayers` is a component so it would be confusing to have it inside another component here. To solve that I've split the line specific settings out into a `LineGizmoConfig` struct and re-used that inside both `GizmoConfig` and the `LineGizmo` component.
- Line joints were not working. Had to re-route some data to fix that.

I also decided to add support for the `Transform` component in this PR after all, since it was easier that I expected.

File: examples/gizmos/3d_gizmos.rs (hunk 3 above)
Comment (pablo-lua) [reply examples/gizmos/3d_gizmos.rs:29]: Yeah, there is already too much in all examples. Maybe it's worth to make a new example just for retained gizmos. But I do agree that it should be make in another PR.

**Review by IceSentry — APPROVED** [review]
I'm not sure how I feel about the naming. I feel like I'd prefer something like `RetainedGizmo` instead of `LineGizmoAsset` or even just `GizmoAsset`. It feels a bit weird that the retained version has a `Line` prefix even if the main difference is the fact that it's retained.

The overall implementation seems good though.

**Comment by tim-blackbird** [comment]
My idea was to expand the `Gizmos` parameter with more than just lines, but it made sense to me to make the retained versions separate? I'm not 100% sure on it

**Comment by IceSentry** [comment]
I think I'd prefer waiting until we actually need them to be separate to show that in the name. Right now it just looks a bit off that there's a Line prefix there.

**Review by tychedelia — APPROVED** [review]
Same minor quibbles about naming, but otherwise looks good. Being able to run gizmos in fixed update is a really compelling use case.

**Comment by tim-blackbird** [comment]
Any thoughts on alternative names for `LineGizmo`?
- `Gizmo`
- `RetainedGizmo`
- `RetainedGizmos`
I don't really like having 'retained' in the name, but besides just `Gizmo` that is all I can think of.

**Review by tychedelia — APPROVED** [review]

**Comment by IceSentry** [comment]
I kinda like `GizmoAsset` because it uses the asset system and it doesn't have the retained naming, but I'm honestly not sure either.

**Comment by tychedelia** [comment]
> I kinda like `GizmoAsset` because it uses the asset system and it doesn't have the retained naming, but I'm honestly not sure either.

We just don't use FooAsset anywhere else, do we?

**Comment by tim-blackbird** [comment]
> We just don't use FooAsset anywhere else, do we?

No, but we do in this PR. We need a name for both the component (currently `LineGizmo`) and the asset (currently `LineGizmoAsset`)

**Comment by pablo-lua** [comment]
> Any thoughts on alternative names for `LineGizmo`?

None, but I think Gizmo is a good idea to follow.

**Comment by tim-blackbird** [comment]
I've thought about it a bit more and decided to go through with removing the 'line' naming from this PR, see the `Rename` commit(dbdb9ad) for the changes

**Comment by alice-i-cecile** [comment]
Bumping to 0.16, sorry. Rendering's stability is too poor to merge this right now.

**Comment by aevyrie** [comment]
This is really nice.

I realize this is a bikeshed, but I'd like to point out that the name Gizmo truly makes no sense any mare. This is just polyline rendering with an immediate and retained API, and helpers to convert primitive shapes into polylines.

**Comment by alice-i-cecile** [comment]
@tim-blackbird I like this and it's well-reviewed. Can you resolve merge conflicts and ping me to merge please?

**Pull request merged**